File: book_examples/bubble_sort/bubble_sort.py
from typing import List


def bubble_sort(arr: List[int]):
    size = len(arr)

    for i in range(size):
        sorted = True
        for j in range(0, size - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                sorted = False
        if sorted:
            break
    print(arr)
    return arr


def test_bubble_sort():
    assert bubble_sort([1, 2, 3, 4, 5]) == [1, 2, 3, 4, 5]
    assert bubble_sort([5, 4, 3, 2, 1]) == [1, 2, 3, 4, 5]
    assert bubble_sort([1, 3, 2, 5, 4]) == [1, 2, 3, 4, 5]
    assert bubble_sort([1, 3, 2, 5, 4]) == [1, 2, 3, 4, 5]


if __name__ == "__main__":
    test_bubble_sort()

File: book_examples/bubble_sort/bubble_sort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

void bubble_sort(std::vector<int>& a) {
    int n = a.size();
    for (int i = 0; i < n - 1; ++i) {
        bool sorted = true;
        for (int j = 0; j < n - 1 - i; ++j) {
            if (a[j] > a[j + 1]) {
                std::swap(a[j], a[j + 1]);
                sorted = false;
            }
        }
        if (sorted) {
            break;
        }
    }
}

int main() {
    int n;

    std::cin >> n;

    std::vector<int> arr(n);

    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }
    bubble_sort(arr);
    for (int i = 0; i < n; ++i) {
        std::cout << arr[i] << " ";
    }
    std::cout << '\n';
    return 0;
}